x-pack/plugins/security_solution/server/lib/entity_analytics/entity_store/elasticsearch_assets/ingest_pipeline.ts

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+import type { IngestProcessorContainer } from '@elastic/elasticsearch/lib/api/types';
+import type { EntityDefinition } from '@kbn/entities-schema';
+import { type FieldRetentionDefinition } from '../field_retention_definition';
+import {
+  debugDeepCopyContextStep,
+  getDotExpanderSteps,
+  getRemoveEmptyFieldSteps,
+  removeEntityDefinitionFieldsStep,
+  retentionDefinitionToIngestProcessorSteps,
+} from './ingest_processor_steps';
+import { getIdentityFieldForEntityType } from '../utils';
+import { getFieldRetentionEnrichPolicyName } from './enrich_policy';
+import type { UnitedEntityDefinition } from '../united_entity_definitions';
+
+const getPlatformPipelineId = (definition: EntityDefinition) => {
+  return `${definition.id}-latest@platform`;
+};
+
+// the field that the enrich processor writes to
+export const ENRICH_FIELD = 'historical';
+
+/**
+ * Builds the ingest pipeline for the field retention policy.
+ * Broadly the pipeline enriches the entity with the field retention enrich policy,
+ * then applies the field retention policy to the entity fields, and finally removes
+ * the enrich field and any empty fields.
+ *
+ * While developing, be sure to set debugMode to true this will keep the enrich field
+ * and the context field in the document to help with debugging.
+ */
+const buildIngestPipeline = ({
+  version,
+  fieldRetentionDefinition,
+  allEntityFields,
+  debugMode,
+  namespace,
+}: {
+  fieldRetentionDefinition: FieldRetentionDefinition;
+  allEntityFields: string[];
+  debugMode?: boolean;
+  namespace: string;
+  version: string;
+}): IngestProcessorContainer[] => {
+  const { entityType, matchField } = fieldRetentionDefinition;
+  const enrichPolicyName = getFieldRetentionEnrichPolicyName({
+    namespace,
+    entityType,
+    version,
+  });
+  return [
+    ...(debugMode ? [debugDeepCopyContextStep()] : []),
+    {
+      enrich: {
+        policy_name: enrichPolicyName,
+        field: matchField,
+        target_field: ENRICH_FIELD,
+      },
+    },
+    {
+      set: {
+        field: '@timestamp',
+        value: '{{entity.lastSeenTimestamp}}',
+      },
+    },
+    {
+      set: {
+        field: 'entity.name',
+        value: `{{${getIdentityFieldForEntityType(entityType)}}}`,
+      },
+    },
+    ...getDotExpanderSteps(allEntityFields),
+    ...retentionDefinitionToIngestProcessorSteps(fieldRetentionDefinition, {
+      enrichField: ENRICH_FIELD,
+    }),
+    ...getRemoveEmptyFieldSteps([...allEntityFields, 'asset', `${entityType}.risk`]),
```

Comment:
what about criticality?

Reply:
Thats handled by allEntityFields, this then cleans up the empty asset object once asset.criticality has been removed